dapp/src/node/mod.rs

```diff
@@ -0,0 +1,176 @@
+struct State<E> {
+    evm: Arc<RwLock<E>>,
+    sender: Wallet<SigningKey>,
+    blocks: Vec<Block<TxHash>>,
+    txs: HashMap<TxHash, Transaction>,
+}
+
+type SharedState<E> = Arc<RwLock<State<E>>>;
+
+async fn handler<E, S>(
+    request: Result<Json<JsonRpcRequest>, JsonRejection>,
+    Extension(state): Extension<SharedState<E>>,
+) -> JsonRpcResponse
+where
+    E: Evm<S>,
+{
+    match request {
+        Err(_) => Error::INVALID_REQUEST.into(),
+        Ok(Json(payload)) => {
+            match serde_json::from_str::<EthRequest>(
+                &serde_json::to_string(&payload)
+                    .expect("deserialized payload should be serializable"),
+            ) {
+                Ok(msg) => {
+                    JsonRpcResponse::new(payload.id(), ResponseContent::success(handle(state, msg)))
+                }
+                Err(e) => {
+                    if e.to_string().contains("unknown variant") {
+                        JsonRpcResponse::new(
+                            payload.id(),
+                            ResponseContent::error(Error::METHOD_NOT_FOUND),
+                        )
+                    } else {
+                        JsonRpcResponse::new(
+                            payload.id(),
+                            ResponseContent::error(Error::INVALID_PARAMS),
+                        )
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn handle<S, E: Evm<S>>(state: SharedState<E>, msg: EthRequest) -> EthResponse {
+    match msg {
+        EthRequest::EthGetBalance(account, _block) => {
+            let balance = state.read().unwrap().evm.read().unwrap().get_balance(account);
+            EthResponse::EthGetBalance(balance)
+        }
+        EthRequest::EthGetTransactionByHash(_tx_hash) => {
+            todo!();
+        }
+        EthRequest::EthSendTransaction(tx) => {
+            let from = if let Some(from) = tx.from() {
+                if state.read().unwrap().sender.address().ne(from) {
+                    unimplemented!("handle: tx.from != node.sender");
+                } else {
+                    *from
+                }
+            } else {
+                state.read().unwrap().sender.address()
+            };
+            let value = *tx.value().unwrap_or(&U256::zero());
+            let calldata = match tx.data() {
+                Some(data) => data.to_vec(),
+                None => vec![],
+            };
+
+            match tx.to() {
+                Some(to) => {
+                    // send tx
+                    let to = match to {
+                        NameOrAddress::Address(addr) => *addr,
+                        NameOrAddress::Name(_) => unimplemented!("handle: tx.to is an ENS name"),
+                    };
+                    // FIXME(rohit): state (and node) must need the chainID
+                    let tx_hash = keccak256(
+                        tx.rlp_signed(1, &state.read().unwrap().sender.sign_transaction_sync(&tx)),
```

Comment:
Should probably make `sender` an arbitrary `S: Signer` if possible? I guess this is bc of the `sync`